### PR TITLE
Ports Dynamic Gamemode Code from Beestation.

### PR DESCRIPTION
**Pull request by Xoxeyos**
Ports over the Dynamic Gamemode Code from Beestation, I have tested it, all seems to be well. Some gamemodes are missing from Dynamic, such as gangs, but I could add that in later. Also a readme for documentation. The readme is in 'code\game\gamemodes\dynamic' .

🆑 Xoxeyos
rscadd: Ports over general dynamic code from Beestation. Thank you qwertyquerty, ShiftyRail, and Toxici11i for making this possible.
/🆑

### CODE REVIEW

**Comment by Xoxeyos** [comment]
Take a look here for some of what it does. https://github.com/tgstation/tgstation/pull/44639

Also here for some images. https://imgur.com/a/KLqOdAY

Here is the Beestation PR, minor edits were needed to be made. https://github.com/BeeStation/BeeStation-Hornet/pull/399/files\

Here's /TG/s explaination on it. https://tgstation13.org/wiki/Game_Mode#Dynamic

Here's /VG/s explanation on the math behind it. However it may have very minor changes from our own. 
https://github.com/vgstation-coders/vgstation13/pull/22410

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+		if(antag_flag_override)
+			if(!(antag_flag_override in M.client.prefs.be_special) || is_banned_from(M.ckey, list(antag_flag_override, ROLE_SYNDICATE)))
+				candidates.Remove(M)
+				continue
+		else
+			if(!(antag_flag in M.client.prefs.be_special) || is_banned_from(M.ckey, list(antag_flag, ROLE_SYNDICATE)))
+				candidates.Remove(M)
+				continue
+		if (M.mind)
+			if (restrict_ghost_roles && M.mind.assigned_role in GLOB.exp_specialmap[EXP_TYPE_SPECIAL]) // Are they playing a ghost role?
```
Comment (SomeguyManperson) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]:
```suggestion
			if (restrict_ghost_roles && (M.mind.assigned_role in GLOB.exp_specialmap[EXP_TYPE_SPECIAL])) // Are they playing a ghost role?

```
Should fix the problem

Comment (Xoxeyos) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: IT WORKS WOOHOO!

**Comment by ThatLing** [comment]
> Some gamemodes are missing from Dynamic, such as gangs, but I could add that in later.

Why not add it now?

**Comment by Hopekz** [comment]
Why is this better?

**Comment by Xoxeyos** [comment]
> 
> 
> Why is this better?

Please put more of an explanation into what you're asking.

**Comment by Xoxeyos** [comment]
> 
> 
> > Some gamemodes are missing from Dynamic, such as gangs, but I could add that in later.
> 
> Why not add it now?

I want to make sure this is well received, and has a test merge, before I add anything else in, to ensure I don't break anything.

**Comment by Hopekz** [comment]
> Please put more of an explanation into what you're asking.

What does this provide over what we already have in terms of the experience?
"Why should this be merged?" a.k.a "Why is this better?"

**Comment by ghost** [comment]
It should be merged because it's Dynamic Hopekz you sack of fuckwits

And the pr's from fucking TG, not bee

**Comment by Xoxeyos** [comment]
> 
> 
> > Please put more of an explanation into what you're asking.
> 
> What does this provide over what we already have in terms of the experience?
> "Why should this be merged?" a.k.a "Why is this better?"

Well, I believe that this will be a fun and interesting gamemode, as it is easily configurable, using the administration game panel, and it will allow the community to give ideas on the subject, and hopefully give the community a good experience. It's also possible to add or remove antagonists, and side antagonists from the rolling list.

**Comment by Xoxeyos** [comment]
> 
> 
> It should be merged because it's Dynamic Hopekz you sack of fuckwits
> 
> And the pr's from fucking TG, not bee

I used bee's code, as /TG/'s code has been heavily modified because of the direction of their codebase, bee is now more similar to our codebase. If I had used /TG/ code, I would have to edit several more files, as, for example, their mob.dm file has different variable names, unlike ours.

**Comment by Hopekz** [comment]
> It should be merged because it's Dynamic Hopekz you sack of fuckwits

Compelling.

> And the pr's from fucking TG, not bee

Better than your usual shitposts I'm awarding you a golden star
download

**Review by fluffe9911 — APPROVED** [review]
Doesn't look like its gonna crash the server or anything

**Comment by Xoxeyos** [comment]
> 
> 
> Doesn't look like its gonna crash the server or anything

HOORAY!

**Comment by swissloaf** [comment]
Make the centcomm roundstart report that generates on paper into an automated announcement, similar to code blue ert, depending on the threat level

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+		if(80 to 99)
+			. += "<b>Black Orbit</b></center><BR>"
+			. += "As part of a mandatory security protocol, we are required to inform you that as a result of your orbital pattern directly behind an astrological body (oriented from our nearest observatory), your station will be under decreased monitoring and support. It is anticipated that your extreme location and decreased surveillance could pose security risks. Avoid unnecessary risks and attempt to keep your station in one piece."
+		if(100)
+			. += "<b>Impending Doom</b></center><BR>"
+			. += "Your station is somehow in the middle of hostile territory, in clear view of any enemy of the corporation. Your likelihood to survive is low, and station destruction is expected and almost inevitable. Secure any sensitive material and neutralize any enemy you will come across. It is important that you at least try to maintain the station.<BR>"
```
Comment (swissloaf) [inline code/game/gamemodes/dynamic/dynamic.dm]: Make the centcomm roundstart report that generates on paper into an automated announcement, similar to code blue ert, depending on the threat level

**Comment by boodaliboo** [comment]
this is fantastic to play thanks xoxo

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 2 above)
Comment (Xoxeyos) [reply code/game/gamemodes/dynamic/dynamic.dm]: I really have no idea on how to do this, but I'll try and find the code blue ERT announcement, and put that in, in some way.

**Comment by boodaliboo** [comment]
does this have yogs exclusive antags?

**Comment by MegaEmpirical** [comment]
Two issues:

1) Blood brothers becoming rev and then being deconverted doesn't show them as antag on admin hud (no BB heart).

2) If revs finish their objective the round ends preventing other antags from doing theirs, such as still active Blood brothers.

Idea: Maybe have antags non-conversion antags start with a stealthed mind shield implant? Or 'special' training that lets them resist.

**Comment by boodaliboo** [comment]
pr of the year

**Comment by boodaliboo** [comment]
dynamic shouldnt have blob spawned through it keep it as a normal midround

**Comment by Xoxeyos** [comment]
> 
> 
> dynamic shouldnt have blob spawned through it keep it as a normal midround

I would say blame Fluffe, but he has already apologized for his sins.

**Comment by Xoxeyos** [comment]
> 
> 
> Two issues:
> 
>     1. Blood brothers becoming rev and then being deconverted doesn't show them as antag on admin hud (no BB heart).
> 
>     2. If revs finish their objective the round ends preventing other antags from doing theirs, such as still active Blood brothers.
> 
> 
> Idea: Maybe have antags non-conversion antags start with a stealthed mind shield implant? Or 'special' training that lets them resist.

This is what I'll work on next, I'm not a good coder of any sort, and I know little about it, so it may take a while.

**Comment by MegaEmpirical** [comment]
I think the best way to do it and possibly the easiest, is to do the 'special training' bit. Look at how the chaplain resist cult conversion, and do it similar to that.

**Comment by Xoxeyos** [comment]
> 
> 
> I think the best way to do it and possibly the easiest, is to do the 'special training' bit. Look at how the chaplain resist cult conversion, and do it similar to that.

Alright, I'll look into that

**Comment by boodaliboo** [comment]
> > I think the best way to do it and possibly the easiest, is to do the 'special training' bit. Look at how the chaplain resist cult conversion, and do it similar to that.
> 
> Alright, I'll look into that

or something along the lines of sec spawning with mindshields is also a good place to look

**Comment by MegaEmpirical** [comment]
I moved away from that idea as I didn't want it to be conformable that someone was an antagonists because of it and I didn't want it to run into any issues with mind shields and their HUDs.

Plus then you can't have it be removable. Which would then lead to revs potentially looking to capture and convert changelings for example (which, admittedly is kind of cool but totally broken).

**Comment by Xoxeyos** [comment]
Maybe it could be added as a traitor item, or something similar.

**Comment by Xoxeyos** [comment]
more random than ever

**Comment by Xoxeyos** [comment]
blah blah blah math math math https://docs.google.com/spreadsheets/d/1kbYWTjYOPS_FrG5fURSHF8e531r3fUQHcyNqfjUoH0k/edit#gid=0

**Comment by Xoxeyos** [comment]
Adding some more stuff to vampire.

**Comment by Xoxeyos** [comment]
Adding some more shadowling stuff soon.

**Comment by Xoxeyos** [comment]
I'm ready for another test merge, or a merge.

**Comment by Marmio64** [comment]
why does it say vampire costs a whole fucking 25 points, but apparently lings and traitors are only 10?

**Comment by MegaEmpirical** [comment]
Vampires can reproduce? Could be why.

**Pull request merged**